File: src/main/java/de/tarent/akka/java/remoteAroundTheWorld/client/PingActor.java
package de.tarent.akka.java.remoteAroundTheWorld.client;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import de.tarent.akka.java.remoteAroundTheWorld.model.Ping;
import de.tarent.akka.java.remoteAroundTheWorld.model.Pong;
import de.tarent.akka.java.remoteAroundTheWorld.model.WakeUp;

class PingActor extends UntypedActor {
    private final LoggingAdapter log = Logging.getLogger(getContext().system(), this);

    private ActorRef pongActor;

    public PingActor(ActorRef pongActor) {
        this.pongActor = pongActor;
    }

    @Override
    public void onReceive(Object message) throws Throwable {
        if(Pong.match(message)) {
            log.info("Receiving a pong!");

        } else if(WakeUp.match(message)) {
            log.info("Ou what!? ... Sending a ping!");
            pongActor.tell(new Ping(), self());

        } else {
            unhandled(message);
        }
    }

    public static Props props(ActorRef pongActor) {
        return Props.create(PingActor.class, pongActor);
    }
}

File: src/main/scala/de/tarent/scala/pingpong/PingActor.scala
package de.tarent.scala.pingpong

import akka.actor.{Actor, ActorRef, Props}
import akka.event.Logging

class PingActor extends Actor {

  val log = Logging(context.system, this)

  val pongActor: ActorRef = context.actorOf(PongActor.props)

  override def receive: Receive = {
    case WakeUp =>
      log.info("WakingUp. Will send a ping.")
      pongActor ! Ping

    case Pong =>
      log.info("Receiving a pong")
  }
}

object PingActor {
  def props: Props = Props[PingActor]
}
